api/stream.go

```diff
@@ -0,0 +1,141 @@
+package api
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/onflow/flow-evm-gateway/config"
+	"github.com/onflow/flow-evm-gateway/storage"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/access/state_stream/backend"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/rs/zerolog"
+)
+
+type StreamAPI struct {
+	logger            zerolog.Logger
+	config            *config.Config
+	blocks            storage.BlockIndexer
+	transactions      storage.TransactionIndexer
+	receipts          storage.ReceiptIndexer
+	accounts          storage.AccountIndexer
+	blocksBroadcaster *engine.Broadcaster
+}
+
+func NewStreamAPI(
+	logger zerolog.Logger,
+	config *config.Config,
+	blocks storage.BlockIndexer,
+	transactions storage.TransactionIndexer,
+	receipts storage.ReceiptIndexer,
+	accounts storage.AccountIndexer,
+	blocksBroadcaster *engine.Broadcaster,
+) *StreamAPI {
+	return &StreamAPI{
+		logger:            logger,
+		config:            config,
+		blocks:            blocks,
+		transactions:      transactions,
+		receipts:          receipts,
+		accounts:          accounts,
+		blocksBroadcaster: blocksBroadcaster,
+	}
+}
+
+// newSubscription creates a new subscription for receiving events from the broadcaster.
+// The data adapter function is used to transform the raw data received from the broadcaster
+// to comply with requested RPC API response schema.
+func (s *StreamAPI) newSubscription(
+	ctx context.Context,
+	broadcaster *engine.Broadcaster,
+	dataAdapter func(any) (any, error),
+) (*rpc.Subscription, error) {
+	notifier, supported := rpc.NotifierFromContext(ctx)
+	if !supported {
+		return &rpc.Subscription{}, rpc.ErrNotificationsUnsupported
+	}
+
+	height, err := s.blocks.LatestEVMHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	sub := backend.NewHeightBasedSubscription(
+		1,
```

Comment:
That `1` looks a bit magic :sweat_smile: . Maybe move it to a variable, a good name might help to understand what it is.

Reply:
Agree yeah, good catch